### ISSUE TITLE
Query: AQL: missing variable #45 (44) for node #8 (CalculationNode) while planning registers (exception location: /work/ArangoDB/arangod/Aql/RegisterPlan.cpp:176)

### ISSUE THREAD

**Issue by ChristyG** [arangodb/arangodb#15097]
## My Environment

* __ArangoDB Version__:         3.7.2 and 3.8.1
* __Storage Engine__:              RocksDB
* __Deployment Mode__:      Single Server 
* __Deployment Strategy__:  Manual Start
* __Infrastructure__:              own 
* __Operating System__:      Ubuntu 20.04 

## Component, Query & Data

__Affected feature__:
AQL query using web interface

__AQL query (if applicable)__:
Simplified query 

 FOR p IN Record
     FILTER p._deleted == false
  
    LET data = []
  SORT TO_NUMBER(SUBSTRING(p.privacyRecordNumber,1))  ASC
  LIMIT 0, 50

LET dpiaRisks = MERGE(
  LET retentionRisk = []
  LET recordRisk = (
    FOR s,e in 1..1 OUTBOUND p HAS_SIMPLE_MASTER_DATA
    COLLECT severity = e.severity WITH COUNT INTO severityCount
    RETURN {severity,severityCount})
    
  FOR s in  append(retentionRisk,recordRisk)
  COLLECT severity = s.severity
  aggregate sumRisk = sum(s.severityCount)
  RETURN {[severity]:sumRisk})
      
RETURN MERGE(p,  {dpiaRisks})

     

## Steps to reproduce
Executing or explain above query result in error

Query: AQL: missing variable #45 (44) for node #8 (CalculationNode) while planning registers (exception location: /work/ArangoDB/arangod/Aql/RegisterPlan.cpp:176). Please report this error to arangodb.com (while optimizing plan) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:290). Please report this error to arangodb.com

Explain: missing variable #45 (44) for node #8 (CalculationNode) while planning registers (exception location: /work/ArangoDB/arangod/Aql/RegisterPlan.cpp:176). Please report this error to arangodb.com (while finalizing)

__Problem__:
Above query result in error. 
If sort and limit is interchanged issue wont occur,
if   FILTER p._deleted == false is removed issue wont occur similarly removal of dpiaRisks.
Working if  FILTER p._deleted == false  is added after dpiaRisks calculation. 

In arango 3.7.2 removing following optimizer resolved the issue

optimizer-rules = -splice-subqueries
optimizer-rules = -optimize-subqueries
optimizer-rules = -optimize-traversals
optimizer-rules = -move-calculations-down
optimizer-rules = -remove-unnecessary-calculations

where as in 3.8.1 issue not resolved even after removal. 

__Expected result__:

Query should work fine irrespective of order of statement

**Comment by jsteemann** [comment]
@ChristyG : I just tried the query with 3.7.15, 3.8.0 and 3.8.2 and I can run `db._explain(query);` just fine with these versions.
Here is the output from my 3.8.0, for example:
```
Query String (559 chars, cacheable: true):
  FOR p IN Record      FILTER p._deleted == false        LET data = []  SORT 
 TO_NUMBER(SUBSTRING(p.privacyRecordNumber,1))  ASC   LIMIT 0, 50  LET dpiaRisks = MERGE(   LET 
 retentionRisk = []   LET recordRisk = (     FOR s,e in 1..1 OUTBOUND p HAS_SIMPLE_MASTER_DATA     
 COLLECT severity = e.severity WITH COUNT INTO severityCount     RETURN {severity,severityCount})     
    FOR s in  append(retentionRisk,recordRisk)   COLLECT severity = s.severity   aggregate sumRisk = 
 sum(s.severityCount)   RETURN {[severity]:sumRisk})        RETURN MERGE(p,  {dpiaRisks}) 

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
 31   IndexNode              0     - FOR p IN Record   /* persistent index scan */    
  6   CalculationNode        0       - LET #18 = TO_NUMBER(SUBSTRING(p.`privacyRecordNumber`, 1))   /* simple expression */   /* collections used: p : Record */
  7   SortNode               0       - SORT #18 ASC   /* sorting strategy: standard */
  8   LimitNode              0       - LIMIT 0, 50
 32   SubqueryStartNode      0       - LET #13 = ( /* subquery begin */
 34   SubqueryStartNode      0         - LET recordRisk = ( /* subquery begin */
 12   TraversalNode          1           - FOR /* vertex optimized away */, e  /* edge */ IN 1..1  /* min..maxPathDepth */ OUTBOUND p /* startnode */  HAS_SIMPLE_MASTER_DATA
 13   CalculationNode        1             - LET #22 = e.`severity`   /* attribute expression */
 14   CollectNode            1             - COLLECT severity = #22 AGGREGATE severityCount = LENGTH()   /* hash */
 30   SortNode               1             - SORT severity ASC   /* sorting strategy: standard */
 15   CalculationNode        1             - LET #24 = { "severity" : severity, "severityCount" : severityCount }   /* simple expression */
 35   SubqueryEndNode        0             - RETURN  #24 ) /* subquery end */
 18   CalculationNode        0         - LET #26 = APPEND([ ], recordRisk)   /* simple expression */
 19   EnumerateListNode      0         - FOR s IN #26   /* list iteration */
 20   CalculationNode        0           - LET #28 = s.`severity`   /* attribute expression */
 21   CalculationNode        0           - LET #30 = s.`severityCount`   /* attribute expression */
 22   CollectNode            0           - COLLECT severity = #28 AGGREGATE sumRisk = SUM(#30)   /* hash */
 29   SortNode               0           - SORT severity ASC   /* sorting strategy: standard */
 23   CalculationNode        0           - LET #32 = { [ severity ] : sumRisk }   /* simple expression */
 33   SubqueryEndNode        0           - RETURN  #32 ) /* subquery end */
 26   CalculationNode        0       - LET dpiaRisks = MERGE(#13)   /* simple expression */
 27   CalculationNode        0       - LET #34 = MERGE(p, { "dpiaRisks" : dpiaRisks })   /* simple expression */   /* collections used: p : Record */
 28   ReturnNode             0       - RETURN #34

Indexes used:
 By   Name                      Type         Collection               Unique   Sparse   Selectivity   Fields           Ranges
 31   idx_1716228643621961728   persistent   Record                   false    false       100.00 %   [ `_deleted` ]   (p.`_deleted` == false)
 12   edge                      edge         HAS_SIMPLE_MASTER_DATA   false    false       100.00 %   [ `_from` ]      base OUTBOUND

Functions used:
 Name        Deterministic   Cacheable   Uses V8
 SUBSTRING   true            true        false  
 TO_NUMBER   true            true        false  
 APPEND      true            true        false  
 MERGE       true            true        false  

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections        Options                                  Filter / Prune Conditions
 12  1..1                       HAS_SIMPLE_MASTER_DATA  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   remove-redundant-calculations
  4   remove-unnecessary-calculations
  5   move-calculations-up-2
  6   move-filters-up-2
  7   use-indexes
  8   remove-filter-covered-by-index
  9   optimize-traversals
 10   remove-unnecessary-calculations-2
 11   move-calculations-down
 12   splice-subqueries
```
However, the "optimize-subqueries" rule is not triggered for me. You mentioned that it may have had an effect.

I assumed there is an index present on `Record`'s `_deleted` attribute. But the query works for me with or without the index.

I am wondering what could be the difference in our environments. Which indexes are present in yours for the underlying collections?

**Comment by ChristyGKrishnan** [comment]
Following index is causing the problem, when I removed that index query is working fine
Index type: skiplist
fields : _deleted, recordType, privacyRecordNumber

Attaching the screenshot of indexes 
![image](https://user-images.githubusercontent.com/50352786/141727431-a3b241b6-72d9-4132-93e0-f2b632837fdc.png)

**Comment by jsteemann** [comment]
@ChristyG : ok, thanks, that helped me reproduce it! Looking into it now.

**Comment by jsteemann** [comment]
Copy&paste code for reproducing:
```js
db = require("@arangodb").db;
db._createEdgeCollection("HAS_SIMPLE_MASTER_DATA");
db._create("Record");
db.Record.ensureIndex({ type: "skiplist", fields: ["_deleted", "recordType", "privacyRecordNumber" ]});
q = `FOR p IN Record FILTER p._deleted == false  LET data = []  SORT TO_NUMBER(SUBSTRING(p.privacyRecordNumber,1)) ASC LIMIT 0, 50  LET dpiaRisks = MERGE( LET retentionRisk = [] LET recordRisk = ( FOR s,e in 1..1 OUTBOUND p HAS_SIMPLE_MASTER_DATA COLLECT severity = e.severity WITH COUNT INTO severityCount RETURN {severity,severityCount})  FOR s in append(retentionRisk,recordRisk) COLLECT severity = s.severity aggregate sumRisk = sum(s.severityCount) RETURN {[severity]:sumRisk})  RETURN MERGE(p, {dpiaRisks})`; 
db._explain(q);
```

**Comment by jsteemann** [comment]
Hi @ChristyG, it just turned out that this error is a duplicate of #14819, which was fixed in version 3.8.2.
I will also open a PR to address the issue in the 3.7 series.

**Comment by maxkernbach** [comment]
Hi @ChristyG,

I am closing this issue since a fix is available in latest versions 3.7.17, 3.8.6 and 3.9.0.
In case an error still occurs, please comment or open a new issue.
